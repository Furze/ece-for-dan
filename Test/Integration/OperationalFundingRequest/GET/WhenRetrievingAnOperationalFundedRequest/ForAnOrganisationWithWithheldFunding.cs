using System;
using System.Collections.Generic;
using System.Linq;
using Bard;
using MoE.ECE.CLI.Data;
using MoE.ECE.Domain.Model.OperationalFunding;
using MoE.ECE.Domain.Read.Model.OperationalFunding;
using MoE.ECE.Integration.Tests.Chapter;
using MoE.ECE.Integration.Tests.Infrastructure;
using Shouldly;
using Xunit;
using Xunit.Abstractions;

namespace MoE.ECE.Integration.Tests.OperationalFundingRequest.GET.WhenRetrievingAnOperationalFundedRequest
{
    public class ForAnOrganisationWithWithheldFunding : IntegrationTestBase<ECEStoryBook, ECEStoryData>
    {
        private Guid _businessEntityId;

        public ForAnOrganisationWithWithheldFunding(RunOnceBeforeAllTests testSetUp, ITestOutputHelper output,
            TestState<ECEStoryBook, ECEStoryData> testState) : base(testSetUp, output, testState)
        {
        }

        protected override void Arrange() =>
            Given
                .A_rs7_skeleton_has_been_created(rs7 =>
                    rs7.OrganisationId = ReferenceData.EceServices.FoxtonPlaycentre.RefOrganisationId)
                .An_rs7_is_ready_for_internal_ministry_review(rs7 =>
                {
                    if (rs7.AdvanceMonths != null)
                    {
                        rs7.AdvanceMonths[0].AllDay = 0;
                        rs7.AdvanceMonths[1].AllDay = 0;
                        rs7.AdvanceMonths[2].AllDay = 0;
                        rs7.AdvanceMonths[3].AllDay = 0;
                    }
                })
                .GetResult(data => _businessEntityId = data.Rs7Model.BusinessEntityId.GetValueOrDefault());

        protected override void Act() =>
            When.Get(
                $"api/operational-funding-requests?business-entity-id={_businessEntityId}&revision=1");

        [Fact]
        public void ThenTheResponseShouldBeOk() => Then.Response.ShouldBe.Ok();
 
        [Fact]
        public void Then_the_business_exceptions_should_not_be_empty()
        {
            var response = Then.Response.Content<ICollection<OperationalFundingRequestModel>>();
                
            response.First().BusinessExceptions.ShouldNotBeEmpty();
            
            response.First().BusinessExceptions?.ShouldContain(model => model.Key == FundingWithheldBusinessException.FundingWithheldKey);
        }

        [Fact]
        public void Then_the_response_snapshot_should_be_ok() =>
            Then.Snapshot()
                .Match<ICollection<OperationalFundingRequestModel>>(IgnoreFieldsFor
                    .CollectionOperationalFundingRequestModel);
    }
}