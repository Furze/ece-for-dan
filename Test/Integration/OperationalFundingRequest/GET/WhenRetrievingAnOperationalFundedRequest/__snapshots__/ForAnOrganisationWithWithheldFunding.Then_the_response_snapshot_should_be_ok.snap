[
  {
    "EntitlementMonths": [
      {
        "MonthName": "February 2020",
        "MonthNumber": 2,
        "Year": 2020,
        "AllDayCertificatedTeacherHours": 0,
        "AllDayNonCertificatedTeacherHours": 0,
        "SessionalCertificatedTeacherHours": 8,
        "SessionalNonCertificatedTeacherHours": 16,
        "TotalWorkingDays": 8,
        "WashUpPlusTen": 0.0,
        "WashUpUnderTwo": 0.0,
        "WashUpTwoAndOver": 0.0,
        "WashUpTwentyHours": 0.0,
        "TotalEntitlement": 0.0,
        "TotalFundsAdvanced": 0.0,
        "TotalEntitlementUnderTwo": 0.0,
        "TotalEntitlementTwoAndOver": 0.0,
        "TotalEntitlementTwentyHours": 0.0,
        "TotalEntitlementPlusTen": 0.0,
        "TotalWashUp": 0.0,
        "FundingComponents": [
          {
            "SessionType": "Sessional",
            "FundingComponentType": "UnderTwo",
            "StartDate": "2020-02-06T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 144,
            "OperatingDays": 8
          },
          {
            "SessionType": "Sessional",
            "FundingComponentType": "TwoAndOver",
            "StartDate": "2020-02-06T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 8,
            "OperatingDays": 8
          },
          {
            "SessionType": "Sessional",
            "FundingComponentType": "PlusTen",
            "StartDate": "2020-02-06T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 184,
            "OperatingDays": 8
          }
        ]
      },
      {
        "MonthName": "March 2020",
        "MonthNumber": 3,
        "Year": 2020,
        "AllDayCertificatedTeacherHours": 0,
        "AllDayNonCertificatedTeacherHours": 0,
        "SessionalCertificatedTeacherHours": 8,
        "SessionalNonCertificatedTeacherHours": 16,
        "TotalWorkingDays": 8,
        "WashUpPlusTen": 0.0,
        "WashUpUnderTwo": 0.0,
        "WashUpTwoAndOver": 0.0,
        "WashUpTwentyHours": 0.0,
        "TotalEntitlement": 0.0,
        "TotalFundsAdvanced": 0.0,
        "TotalEntitlementUnderTwo": 0.0,
        "TotalEntitlementTwoAndOver": 0.0,
        "TotalEntitlementTwentyHours": 0.0,
        "TotalEntitlementPlusTen": 0.0,
        "TotalWashUp": 0.0,
        "FundingComponents": [
          {
            "SessionType": "Sessional",
            "FundingComponentType": "UnderTwo",
            "StartDate": "2020-03-05T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 144,
            "OperatingDays": 8
          },
          {
            "SessionType": "Sessional",
            "FundingComponentType": "TwoAndOver",
            "StartDate": "2020-03-05T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 8,
            "OperatingDays": 8
          },
          {
            "SessionType": "Sessional",
            "FundingComponentType": "PlusTen",
            "StartDate": "2020-03-05T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 184,
            "OperatingDays": 8
          }
        ]
      },
      {
        "MonthName": "April 2020",
        "MonthNumber": 4,
        "Year": 2020,
        "AllDayCertificatedTeacherHours": 0,
        "AllDayNonCertificatedTeacherHours": 0,
        "SessionalCertificatedTeacherHours": 9,
        "SessionalNonCertificatedTeacherHours": 18,
        "TotalWorkingDays": 9,
        "WashUpPlusTen": 0.0,
        "WashUpUnderTwo": 0.0,
        "WashUpTwoAndOver": 0.0,
        "WashUpTwentyHours": 0.0,
        "TotalEntitlement": 0.0,
        "TotalFundsAdvanced": 0.0,
        "TotalEntitlementUnderTwo": 0.0,
        "TotalEntitlementTwoAndOver": 0.0,
        "TotalEntitlementTwentyHours": 0.0,
        "TotalEntitlementPlusTen": 0.0,
        "TotalWashUp": 0.0,
        "FundingComponents": [
          {
            "SessionType": "Sessional",
            "FundingComponentType": "UnderTwo",
            "StartDate": "2020-04-02T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 162,
            "OperatingDays": 9
          },
          {
            "SessionType": "Sessional",
            "FundingComponentType": "TwoAndOver",
            "StartDate": "2020-04-02T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 9,
            "OperatingDays": 9
          },
          {
            "SessionType": "Sessional",
            "FundingComponentType": "PlusTen",
            "StartDate": "2020-04-02T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 207,
            "OperatingDays": 9
          }
        ]
      },
      {
        "MonthName": "May 2020",
        "MonthNumber": 5,
        "Year": 2020,
        "AllDayCertificatedTeacherHours": 0,
        "AllDayNonCertificatedTeacherHours": 0,
        "SessionalCertificatedTeacherHours": 9,
        "SessionalNonCertificatedTeacherHours": 18,
        "TotalWorkingDays": 9,
        "WashUpPlusTen": 0.0,
        "WashUpUnderTwo": 0.0,
        "WashUpTwoAndOver": 0.0,
        "WashUpTwentyHours": 0.0,
        "TotalEntitlement": 0.0,
        "TotalFundsAdvanced": 0.0,
        "TotalEntitlementUnderTwo": 0.0,
        "TotalEntitlementTwoAndOver": 0.0,
        "TotalEntitlementTwentyHours": 0.0,
        "TotalEntitlementPlusTen": 0.0,
        "TotalWashUp": 0.0,
        "FundingComponents": [
          {
            "SessionType": "Sessional",
            "FundingComponentType": "UnderTwo",
            "StartDate": "2020-05-01T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 162,
            "OperatingDays": 9
          },
          {
            "SessionType": "Sessional",
            "FundingComponentType": "TwoAndOver",
            "StartDate": "2020-05-01T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 9,
            "OperatingDays": 9
          },
          {
            "SessionType": "Sessional",
            "FundingComponentType": "PlusTen",
            "StartDate": "2020-05-01T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 207,
            "OperatingDays": 9
          }
        ]
      }
    ],
    "OrganisationId": 10007,
    "BusinessEntityId": "b22238b3-06f1-453d-ad42-9bd7c2608114",
    "RequestId": "141122",
    "FundingPeriodMonth": "July",
    "FundingYear": 2021,
    "RevisionNumber": 1,
    "TotalWashUp": 0.0,
    "TotalAdvance": 0.0,
    "AdvanceMonths": [
      {
        "MonthName": "July 2020",
        "MonthNumber": 7,
        "Year": 2020,
        "ParentLedWorkingDays": null,
        "AllDayWorkingDays": 0,
        "SessionalWorkingDays": 0,
        "AmountPayablePlusTen": 0.0,
        "AmountPayableUnderTwo": 0.0,
        "AmountPayableTwoAndOver": 0.0,
        "AmountPayableTwentyHours": 0.0,
        "TotalDays": 0,
        "TotalAdvance": 0.0,
        "FundingComponents": []
      },
      {
        "MonthName": "August 2020",
        "MonthNumber": 8,
        "Year": 2020,
        "ParentLedWorkingDays": null,
        "AllDayWorkingDays": 0,
        "SessionalWorkingDays": 0,
        "AmountPayablePlusTen": 0.0,
        "AmountPayableUnderTwo": 0.0,
        "AmountPayableTwoAndOver": 0.0,
        "AmountPayableTwentyHours": 0.0,
        "TotalDays": 0,
        "TotalAdvance": 0.0,
        "FundingComponents": []
      },
      {
        "MonthName": "September 2020",
        "MonthNumber": 9,
        "Year": 2020,
        "ParentLedWorkingDays": null,
        "AllDayWorkingDays": 0,
        "SessionalWorkingDays": 0,
        "AmountPayablePlusTen": 0.0,
        "AmountPayableUnderTwo": 0.0,
        "AmountPayableTwoAndOver": 0.0,
        "AmountPayableTwentyHours": 0.0,
        "TotalDays": 0,
        "TotalAdvance": 0.0,
        "FundingComponents": []
      },
      {
        "MonthName": "October 2020",
        "MonthNumber": 10,
        "Year": 2020,
        "ParentLedWorkingDays": null,
        "AllDayWorkingDays": 0,
        "SessionalWorkingDays": 0,
        "AmountPayablePlusTen": 0.0,
        "AmountPayableUnderTwo": 0.0,
        "AmountPayableTwoAndOver": 0.0,
        "AmountPayableTwentyHours": 0.0,
        "TotalDays": 0,
        "TotalAdvance": 0.0,
        "FundingComponents": []
      }
    ],
    "MatchingAdvanceMonths": [],
    "Equity": null,
    "BusinessExceptions": [
      {
        "Key": "FundingWithheld",
        "Description": "This service is not funded"
      }
    ]
  }
]
